lib/services/expiry/expiry.go

```diff
@@ -0,0 +1,250 @@
+			continue
+		}
+
+		if err := s.processRequests(); err != nil {
+			s.Log.WarnContext(s.ctx, "error processing access requests", "error", err)
+		}
+		ctx, cancel := context.WithCancel(lease)
+		defer cancel()
+		lease.Stop()
+		if err := lease.Wait(); err != nil {
+			s.Log.WarnContext(ctx, "error cleaning up semaphore", "error", err)
+		}
+		select {
+		case <-s.ctx.Done():
+			return nil
+		case <-poll.Next():
+		}
+	}
+}
+func (s *Service) processRequests() error {
+	requestsExpired := 0
+	nextPageStart := ""
+	for {
+		var page []*types.AccessRequestV3
+		var err error
+		readTime := s.Clock.Now()
+		page, nextPageStart, err = s.getNextPageOfAccessRequests(nextPageStart)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if len(page) == 0 || requestsExpired >= maxExpiresPerCycle {
+			return nil
+		}
+		minPageDelay := time.After(retryutils.SeventhJitter(minPageDelay))
+		for _, req := range page {
+			if !s.shouldExpire(req, readTime) {
+				continue
+			}
+			if err := s.expireRequest(s.ctx, req); err != nil {
+				return trace.Wrap(err)
+			}
+			requestsExpired++
+			if requestsExpired >= maxExpiresPerCycle {
+				return nil
+			}
+		}
+		<-minPageDelay
+	}
+}
+
+func (s *Service) getNextPageOfAccessRequests(startKey string) ([]*types.AccessRequestV3, string, error) {
+	req := &proto.ListAccessRequestsRequest{
+		Sort:       proto.AccessRequestSort_CREATED,
+		Descending: true,
+		Limit:      accessRequestPageLimit,
+		StartKey:   startKey,
+	}
+	resp, err := s.AccessPoint.ListAccessRequests(s.ctx, req)
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+
+	return resp.AccessRequests, resp.NextKey, nil
+}
+
+func (s *Service) shouldExpire(req types.AccessRequest, readTime time.Time) bool {
+	expires := req.Expiry()
+	if req.GetState() == types.RequestState_PENDING {
+		expires = expires.Add(pendingRequestGracePeriod)
+	}
+	return readTime.After(expires)
+}
+
+func (s *Service) expireRequest(ctx context.Context, req types.AccessRequest) error {
+	var annotations *apievents.Struct
+	if sa := req.GetSystemAnnotations(); len(sa) > 0 {
+		var err error
+		annotations, err = apievents.EncodeMapStrings(sa)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	expiry := req.Expiry()
+	event := &apievents.AccessRequestExpire{
+		Metadata: apievents.Metadata{
+			Type: events.AccessRequestExpireEvent,
+			Code: events.AccessRequestExpireCode,
+		},
+		ResourceMetadata: apievents.ResourceMetadata{
+			Expires: req.GetAccessExpiry(),
+		},
+		Roles:                req.GetRoles(),
+		RequestedResourceIDs: apievents.ResourceIDs(req.GetRequestedResourceIDs()),
+		RequestID:            req.GetName(),
+		RequestState:         req.GetState().String(),
+		Reason:               req.GetRequestReason(),
+		MaxDuration:          req.GetMaxDuration(),
+		Annotations:          annotations,
+		ResourceExpiry:       &expiry,
+	}
+	if err := s.Emitter.EmitAuditEvent(ctx, event); err != nil {
```

Comment:
Leave a comment explaining why we're emitting event before deleting the request.